presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsTracker.java

```diff
@@ -176,21 +194,70 @@ public Map<PlanNodeWithHash, PlanStatisticsWithSourceInfo> getQueryStats(QueryIn
+                            // we're doing a depth-first traversal of the plan tree: cache the final agg so that when we encounter the partial agg we can come back
+                            // and update the partial agg statistics
+                            aggregationNodeMap.put(((AggregationNode) planNode).getAggregationId().get(), new FinalAggregationStatsInfo(planNodeWithHash, planStatsWithSourceInfo));
                         }
-                        planStatistics.put(
-                                planNodeWithHash,
-                                new PlanStatisticsWithSourceInfo(
-                                        planNode.getId(),
-                                        newPlanNodeStats,
-                                        new HistoryBasedSourceInfo(Optional.of(hash), Optional.of(inputTableStatistics))));
                     }
                 }
             }
         }
-        return ImmutableMap.copyOf(planStatistics);
+        return ImmutableMap.copyOf(planStatisticsMap);
+    }
+
+    private static void updatePartialAggregationStatistics(
+            AggregationNode partialAggregationNode,
+            Map<Integer, FinalAggregationStatsInfo> aggregationNodeStats,
+            PartialAggregationStatistics partialAggregationStatistics,
+            Map<PlanNodeWithHash, PlanStatisticsWithSourceInfo> planStatisticsMap)
+    {
+        if (!partialAggregationNode.getAggregationId().isPresent() || !aggregationNodeStats.containsKey(partialAggregationNode.getAggregationId().get())) {
+            return;
+        }
+
+        // find the stats for the matching final aggregation node (the partial and the final node share the same aggregationId)
+        FinalAggregationStatsInfo finalAggregationStatsInfo = aggregationNodeStats.get(partialAggregationNode.getAggregationId().get());
+        PlanStatisticsWithSourceInfo planStatisticsWithSourceInfo = finalAggregationStatsInfo.getPlanStatisticsWithSourceInfo();
+        PlanStatistics planStatisticsFinalAgg = planStatisticsWithSourceInfo.getPlanStatistics();
+        planStatisticsFinalAgg = planStatisticsFinalAgg.updateAggregationStatistics(partialAggregationStatistics);
+
+        planStatisticsMap.put(
+                finalAggregationStatsInfo.getPlanNodeWithHash(),
+                new PlanStatisticsWithSourceInfo(
+                    planStatisticsWithSourceInfo.getId(),
+                    planStatisticsFinalAgg,
+                    planStatisticsWithSourceInfo.getSourceInfo()));
+    }
+
+    private PartialAggregationStatistics constructAggregationNodeStatistics(PlanNode planNode, Map<PlanNodeId, PlanNodeStats> planNodeStatsMap, double outputBytes, double outputPositions)
+    {
+        PlanNode childNode = planNode.getSources().get(0);
+        PlanNodeStats childNodeStats = planNodeStatsMap.get(childNode.getId());
+        if (childNodeStats != null) {
+            double partialAggregationInputBytes = adjustedOutputBytes(childNode, childNodeStats);
```

Comment:
@mlyublena where the adjustedOutputBytes are directly used to populate the PartialAggregationStatistics

Reply:
thanks @feilong-liu , that is indeed the fix, I wanted to investigate the case where this happens so I can test that it is fixed. The problem was when doing a GROUP BY on the partition key, the partition key is not materialized in the  page sent through partial aggregation, so the byte count is 0 (and then becomes negative because we exclude the hash variables). So the problematic queries were of this pattern:
```
select ds from lineitem group by ds;
```

new PR here:
https://github.com/prestodb/presto/pull/21502